### PR TITLE
cmd/libsnap: add sc_string_append_char

### PR DESCRIPTION
**Pull request by zyga**
This patch adds a paranoid-security function for appending a character
to a C string along with complete test suite.

Signed-off-by: Zygmunt Krynicki <zygmunt.krynicki@canonical.com>

### CODE REVIEW

File: cmd/libsnap-confine-private/string-utils.c [hunk 1]
```diff
@@ -99,6 +99,34 @@ size_t sc_string_append(char *dst, size_t dst_size, const char *str)
...
+	}
+	size_t dst_len = strnlen(dst, dst_size);
+	if (dst_len == dst_size) {
+		die("cannot append character: dst is unterminated");
+	}
+	size_t max_str_len = dst_size - dst_len;
+	if (max_str_len < 2) {
+		die("cannot append character: not enough space");
+	}
+	if (c < 0 || c > 255) {
```
Comment (chipaca) [inline cmd/libsnap-confine-private/string-utils.c]: This is puzzling to me for three reasons:

1. `char` is signed in the arches we care about (it's arch-dependent, but I don't know one with unsigned chars). 
2. Why isn't the function itself taking a `char` directly?
3. Are we doing something other than appending ASCII? If not, why not check for the ASCII range, which is 7 bits?

File: cmd/libsnap-confine-private/string-utils.c [hunk 2]
```diff
@@ -99,6 +99,34 @@ size_t sc_string_append(char *dst, size_t dst_size, const char *str)
...
+		die("cannot append character: character out of range");
+	}
+	if (c == 0) {
+		die("cannot append character: cannot append string terminator");
+	}
+	// Append the character and terminate the string.
+	dst[dst_len + 0] = c;
+	dst[dst_len + 1] = '\0';
+	// Return the new size
+	return strlen(dst);
```
Comment (stolowski) [inline cmd/libsnap-confine-private/string-utils.c]: I think the length here could easily be derived, it should be the original length + 1, no need to iterate over the string again to find it out.

File: cmd/libsnap-confine-private/string-utils.c (hunk 1 above)
Comment (chipaca) [inline cmd/libsnap-confine-private/string-utils.c]: Correction: char is unsigned on arm and a ppc6el and such. https://wiki.debian.org/ArchitectureSpecificsMemo

Comment (zyga) [inline cmd/libsnap-confine-private/string-utils.c]: This is now simplified after some discussion on telegram.

**Pull request merged**